torchbenchmark/models/sam/__init__.py

```diff
@@ -47,14 +84,72 @@ def get_module(self):
         return self.model, (example_input, multimask_output)
             
     def train(self):
-        error_msg = """
-            As of May 17, 2023
-            Some base VIT checkpoints are available for SAM but getting the dataset
-            requires a research license. It's easy to make up a training loop on random
-            data and if that's interesting please let @msaroufim know
-            https://github.com/facebookresearch/segment-anything#dataset
-        """
-        return NotImplementedError(error_msg)
+        # Code copied from https://github.com/bowang-lab/MedSAM/blob/main/finetune_and_inference_tutorial.py
+        
+        
+        npz_tr_path = 'data/Npz_files/CT_Abd-Gallbladder/train'
+        work_dir = './work_dir'
+        task_name = 'CT_Abd-Gallbladder'
+        # prepare SAM model
+        model_type = 'vit_b'
+        checkpoint = 'work_dir/SAM/sam_vit_b_01ec64.pth'
+        device = 'cuda:0'
+        model_save_path = os.path.join(work_dir, task_name)
+        os.makedirs(model_save_path, exist_ok=True)
+        self.model.train()
+        optimizer = torch.optim.Adam(self.model.mask_decoder.parameters(), lr=1e-5, weight_decay=0)
+        seg_loss = monai.losses.DiceCELoss(sigmoid=True, squared_pred=True, reduction='mean')
+
+        num_epochs = 100
+        losses = []
+        best_loss = 1e10
+
+        train_dataset = NpzDataset(npz_tr_path)
+        train_dataloader = DataLoader(train_dataset, batch_size=64, shuffle=True)
```

Comment:
We shouldn't involve any DataLoader code in `train()` or `eval()`. The principle is to prefetch just 1 batch of input in `__init__()`, and run it in `train()`.